apps/src/dance/LottieDancerRenderer.ts

```diff
@@ -0,0 +1,673 @@
+      adlibOption?: string;
+      choices?: string[];
+      variant?: number | string;
+    } | null;
+
+    const {adlibOption, choices, variant} = localStorageDancer || {};
+
+    const resolvedDancerName =
+      choices && String(variant)
+        ? `${choices.join('-')}-0${String(variant)}`
+        : TEST_GENERATED_DANCER;
+
+    this.headScale = 0.5;
+    this.assetsPath =
+      (appConfig.getValue('path') as string) || adlibOption || ASSETS_FOLDER;
+    this.skeletonName =
+      (appConfig.getValue('skeleton') as string)?.toLowerCase() ||
+      TEST_BASE_DANCER;
+    this.variant = (variant as string) || '00';
+    this.dancerName =
+      (appConfig.getValue('dancer') as string)?.toLowerCase() ||
+      resolvedDancerName;
+    this.cachedAnimationData = {};
+  }
+
+  /** DanceParty hands us a p5.Graphics 2D context. */
+  init(ctx: Canvas2D) {
+    this.ctx = ctx;
+  }
+
+  async setSource(danceMove?: string | null): Promise<void> {
+    if (!danceMove) {
+      this._destroyAnim();
+      this.animationData = null;
+      this.palette = null;
+      this.totalFrames = null;
+      return;
+    }
+
+    const move = String(danceMove).toLowerCase();
+    let animData: LottieJSON;
+    if (this.cachedAnimationData[move]) {
+      animData = this.cachedAnimationData[move];
+    } else {
+      const jsonUrl = this._resolveAnimationUrl(move);
+      const metadataUrl = this._resolveMetadataUrl();
+
+      animData = await this._fetchJson<LottieJSON>(jsonUrl);
+      const metadataJson = await this._fetchJson<Record<string, unknown>>(
+        metadataUrl
+      );
+
+      this.palette = this._normalizePalette(metadataJson);
+
+      // Recolor assets based on hard-coded accessory-name rules.
+      this._applyColorMapping(animData, this.palette);
+
+      // Replace vector head with an image, when head.png is available.
+      const headInfo = await this._fetchHeadImageInfo();
+      if (headInfo) {
+        const headPre = this._findHeadPrecompLayerDeep(animData);
+        if (headPre?.refId) {
+          const headComp = this._getAssetById(animData, headPre.refId);
+          if (headComp && Array.isArray(headComp.layers)) {
+            const {insertIndex, headKs} = this._hideVectorHeadInComp(headComp);
+            const assetId = this._ensureHeadImageAsset(
+              animData,
+              headInfo.dataUrl,
+              headInfo.w,
+              headInfo.h
+            );
+            this._insertHeadImageLayer(
+              headComp,
+              insertIndex,
+              assetId,
+              headInfo.w,
+              headInfo.h,
+              headKs
+            );
+          }
+        }
+      }
+      // Memoize transformed Lottie JSON per move (in-memory cache) so subsequent setSource calls skip recolor/head work.
+      // This is useful if the same dance move is used later in a song, or if there are multiple generated dancers using the same move.
+      this.cachedAnimationData[move] = animData;
```

Comment:
As part of the profiling work, curious how much memory each of these blobs takes.